hedera-node/hedera-app/src/main/java/com/hedera/node/app/util/ProtobufUtils.java

```diff
@@ -0,0 +1,177 @@
+        return extractFieldBytes(serializedTransaction.toReadableSequentialData(), TransactionSchema.BODY);
+    }
+
+    @NonNull
+    public static List<Bytes> extractInnerTransactionBytes(@NonNull final Bytes serializedAtomicBatchTxn)
+            throws IOException, ParseException {
+        final var atomicTransactionBody = extractFieldBytes(
+                serializedAtomicBatchTxn.toReadableSequentialData(), TransactionBodySchema.ATOMIC_BATCH);
+        return extractRepeatedFieldBytes(
+                atomicTransactionBody.toReadableSequentialData(), AtomicBatchTransactionBodySchema.TRANSACTIONS);
+    }
+
+    @NonNull
+    public static Bytes extractPaymentBytes(@NonNull final Bytes serializedQuery) throws IOException, ParseException {
+        final var queryBody = extractQuery(serializedQuery.toReadableSequentialData());
+        final var queryHeader =
+                extractFieldBytes(queryBody.toReadableSequentialData(), TransactionGetReceiptQuerySchema.HEADER);
+        return extractFieldBytes(queryHeader.toReadableSequentialData(), QueryHeaderSchema.PAYMENT);
+    }
+
+    @NonNull
+    private static Bytes extractFieldBytes(
+            @NonNull final ReadableSequentialData input, @NonNull final FieldDefinition field)
+            throws IOException, ParseException {
+        if (field.repeated()) {
+            throw new IllegalArgumentException("Cannot extract field bytes for a repeated field: " + field);
+        }
+        if (ProtoWriterTools.wireType(field) != ProtoConstants.WIRE_TYPE_DELIMITED) {
+            throw new IllegalArgumentException("Cannot extract field bytes for a non-length-delimited field: " + field);
+        }
+        while (input.hasRemaining()) {
+            final int tag;
+            // hasRemaining() doesn't work very well for streaming data, it returns false only when
+            // the end of input is already reached using a read operation. Let's catch an underflow
+            // (actually, EOF) exception here and exit cleanly. Underflow exception in any other
+            // place means malformed input and should be rethrown
+            try {
+                tag = input.readVarInt(false);
+            } catch (final BufferUnderflowException e) {
+                // No more fields
+                break;
+            }
+            final int fieldNum = tag >> TAG_FIELD_OFFSET;
+            final ProtoConstants wireType = ProtoConstants.get(tag & ProtoConstants.TAG_WIRE_TYPE_MASK);
+            if (fieldNum == field.number()) {
+                if (wireType != ProtoConstants.WIRE_TYPE_DELIMITED) {
+                    throw new ParseException("Unexpected wire type: " + tag);
+                }
+                final int length = input.readVarInt(false);
+                return input.readBytes(length);
+            } else {
+                ProtoParserTools.skipField(input, wireType);
+            }
+        }
+        throw new ParseException("Field not found: " + field);
```

Comment:
PBJ 0.9.18 handles it slightly differently, it returns null if the field is not found rather than throwing an exception. When this code is migrated to use `ProtoParserUtils.extractFieldBytes()` directly, it's something to remember about